openedx/core/lib/tests/assertions/events.py

```diff
@@ -0,0 +1,134 @@
+
+import json
+import pprint
+
+
+def assert_event_matches(expected, actual, tolerate=None):
+    """
+    Compare two event dictionaries.
+
+    Fail if any discrepancies exist, and output the list of all discrepancies. The intent is to produce clearer
+    error messages than "{ some massive dict } != { some other massive dict }", instead enumerating the keys that
+    differ. Produces period separated "paths" to keys in the output, so "context.foo" refers to the following
+    structure:
+
+        {
+            'context': {
+                'foo': 'bar'  # this key, value pair
+            }
+        }
+    """
+    differences = get_event_differences(expected, actual, tolerate=tolerate)
+    if len(differences) > 0:
+        debug_info = [
+            '',
+            'Expected:',
+            block_indent(expected),
+            'Actual:',
+            block_indent(actual),
+            'Tolerating:',
+            block_indent(get_tolerate_or_default(tolerate)),
+        ]
+        differences = ['* ' + d for d in differences]
+        message_lines = differences + debug_info
+        raise AssertionError('Unexpected differences found in structs:\n\n' + '\n'.join(message_lines))
+
+
+def assert_events_equal(expected, actual):
+    """
+    Strict comparison of two events.
+
+    This asserts that every field in the real event exactly matches the expected event.
+    """
+    assert_event_matches(expected, actual, tolerate=frozenset())
+
+
+def get_event_differences(expected, actual, tolerate=None):
+    tolerate = get_tolerate_or_default(tolerate)
+
+    # Some events store their payload in a JSON string instead of a dict. Comparing these strings can be problematic
+    # since the keys may be in different orders, so we parse the string here if we were expecting a dict.
+    if 'string_payload' in tolerate:
+        expected = parse_event_payload(expected.copy())
+        actual = parse_event_payload(actual.copy())
```

Comment:
I think you could move the copy into the parse_event_payload, and only call it when you need to reset the event.  That way you're not copying *every* event passed in here (whether needing it or not).